app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -168,39 +172,119 @@ public function indexAction()
-            'city' => $request_data->city,
-            'region' => $request_data->region_code,
-            'postcode' => $request_data->zip_code,
-            'country_id' => $request_data->country_code
-        );
-        $quote->getShippingAddress()->addData($address_data);
-        $quote->getBillingAddress()->addData($address_data);
+        $shippingAddress = $quote->getShippingAddress();
 
-        try {
-            $estimate_response = Mage::helper('boltpay/api')->getShippingAndTaxEstimate($quote);
-        } catch (Exception $e) {
-            $estimate_response = null;
+        $countItems = $quote->getItemsCollection()->getSize();
+        $countryId = $shippingAddress->getCountryId();
+        if ($shippingAddress && empty($countryId) && $countItems)
+        {
+            $request_json = file_get_contents('php://input');
+            $request_data = json_decode($request_json);
+
+            $address_data = array(
+                'city'          => $request_data->city,
+                'region'        => $request_data->region_code,
+                'region_name'   => $request_data->region_name,
+                'postcode'      => $request_data->zip_code,
+                'country_id'    => $request_data->country_code
+            );
+
+            /** @var Mage_Directory_Model_Country $countryModel */
+            $countryModel = Mage::getModel('directory/country');
+            $countryObj = $countryModel->load($address_data['country_id']);
+            $isRegionAvailable = ($countryObj->getRegionCollection()->getSize() > 0);
+            if (!$isRegionAvailable) {
+                // If country does not have region options for dropdown.
+                $address_data['region'] = $address_data['region_name'];
+            }
+
+            /** @var Mage_Directory_Model_Region $regionModel */
+            $regionModel = Mage::getModel('directory/region');
+            $regionObj= $regionModel->loadByCode($address_data['region'], $address_data['country_id']);
+            $address_data['region_id'] = ($regionObj) ? $regionObj->getId() : null;
+
+            ////////////////////////////////////////////////////////////////////////
+            // Clear previously set estimates.  This helps if this
+            // process fails or is aborted, which will force the actual index action
+            // to get fresh data instead of reading from the session cache
+            ////////////////////////////////////////////////////////////////////////
+            $cacheKeyLocation = $quote->getId() . '_' . $address_data['country_id'];
+            $prefetchCacheKey = $this->generateCacheKey($quote, $address_data);
+            Mage::app()->getCache()->remove('quote_location_' . $cacheKeyLocation);
+            Mage::app()->getCache()->remove($prefetchCacheKey);
+            ////////////////////////////////////////////////////////////////////////
+
+            $quote->getShippingAddress()->addData($address_data);
```

Comment:
Does this address John's concern? It seems we always override address in quote, which we should avoid - if quote already has address info, it shouldn't be overridden by IP-based inferred address
https://app.asana.com/0/436430054069245/633055371214186/f

Reply:
A bit above we have 
if ($shippingAddress && empty($countryId) && $countItems)

IF in quote exist $countryId we do not do a recalculation.

Reply:
maybe I'm a bit paranoid but is there any situation where $countryId is not set but region etc. are set?

Reply:
@daisy1754 I did check a recent database to confirm that country id is set whenever there was an address. I'm not sure if this is always the case, but I think it's a good assumption. No reason we couldn't just add that though to be safe.